wandb/sdk/internal/system/assets/open_metrics.py

```diff
@@ -0,0 +1,176 @@
+import logging
+import multiprocessing as mp
+from collections import defaultdict, deque
+from hashlib import md5
+from types import ModuleType
+from typing import TYPE_CHECKING, Dict, List, Union
+
+import requests
+
+import wandb
+from wandb.sdk.lib import telemetry
+
+from .aggregators import aggregate_last, aggregate_mean
+from .interfaces import Interface, Metric, MetricsMonitor
+
+if TYPE_CHECKING:
+    from typing import Deque, Optional
+
+    from wandb.sdk.internal.settings_static import SettingsStatic
+
+
+logger = logging.getLogger(__name__)
+
+
+prometheus_client_parser: "Optional[ModuleType]" = None
+try:
+    import prometheus_client.parser  # type: ignore
+
+    prometheus_client_parser = prometheus_client.parser
+except ImportError:
+    pass
+
+
+class OpenMetricsMetric:
+    """
+    Container for all the COUNTER and GAUGE metrics extracted from an
+    OpenMetrics endpoint.
+    """
+
+    def __init__(self, name: str, url: str) -> None:
+        self.name = name
+        self.url = url
+        self.session: Optional["requests.Session"] = None
+        self.samples: "Deque[dict]" = deque([])
+        # {"<metric name>": {"<labels hash>": <index>}}
+        self.label_map: "Dict[str, Dict[str, int]]" = defaultdict(dict)
+        # {"<labels hash>": <labels>}
+        self.label_hashes: "Dict[str, dict]" = {}
+
+    def setup(self) -> None:
+        if self.session is None:
+            self.session = requests.Session()
+
+    def teardown(self) -> None:
+        if self.session is not None:
+            self.session.close()
+            self.session = None
+
+    def parse_open_metrics_endpoint(self) -> Dict[str, Union[str, int, float]]:
+        assert prometheus_client_parser is not None
+        assert self.session is not None
+
+        response = self.session.get(self.url)
```

Comment:
are you ok with this not having a timeout?

Reply:
this should not be a possibility anymore